accepted/0000-retail-pxe-database.md

```diff
@@ -0,0 +1,259 @@
+* (re)generate single PXE entry for given pxe server and terminal (use case when changing the kernel option of the terminal, e.g. for debugging purposes)
+
+`saltboot.generatePXEentry(sessionKey, server_id)`
+
+`saltboot.generatePXEentries(sessionKey, pxeserver_id)`
+
+### API to show/list entries
+
+API is recommended to be implemented for sake of complete API based access and it should be possible to:
+
+* list PXE entries for given pxeserver id
+* show PXE entries for given server id
+
+`saltboot.listPXEentries(sessionKey, pxeserver_id)` -> list of `server_id` with pxe entry for given `pxeserver_id`
+
+`saltboot.getPXEentry(sessionKey, server_id)` -> dictionary of stored values for given `server_id`
+
+### API to remove PXE entries
+
+It is up to discussion as the entry should be automatically removed when terminal server is removed from Uyuni/SUSE Manager.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+Why should we **not** do this?
+
+  * obscure corner cases
+  
+    Currently unknown
+  
+  * will it impact performance?
+  
+    Multiple terminals booting at the same time - current state is handled solely by salt, proposed solution adds communication with database - increasing latency and some load on database
+  
+  * what other parts of the product will be affected?
+  
+    There should be no affect to other parts of product
+  
+  * will the solution be hard to maintain in the future?
+  
+    As is the case even now, the code is split across different projects and github repositories. Saltboot code in [uyuni-retail/saltboot-formula](https://github.com/uyuni-project/retail/tree/master/saltboot-formula), `pxe/terminal_entry` state in [salt-formulas/pxe-formula](https://github.com/SUSE/salt-formulas/tree/master/pxe-formula) next adding java event handler adds third repository. However code itself is not complex and this code split can be mitigated either by moving parts closer together (like [integrating saltboot to uyuni](https://github.com/SUSE/spacewalk/issues/10777)) or good code commenting.
+  
+
+# Alternatives
+[alternatives]: #alternatives
+
+- Automatic detection of partition where salt configuration is stored
+
+  Draft PR [is here](https://github.com/uyuni-project/retail/pull/71). This removes the need to reregister terminal in case of lost PXE entry. However this does not preserve terminal-specific kernel command line options and always boot default initrd and kernel.
+
+- TPM as storage for salt configuration
+
+  Trusted Platform Module is required to be available on every hardware certified to run MS Windows 10, essentially every new hardware. It can be used as a secure storage for salt keys and minion id instead of storing it on hard drive. This would solve the same problem as automatic detection of salt configuration with the same drawbacks. This would also need some more time for research and packaging ensuring tpm2-tss is available in saltboot initrd.
```

Comment:
Given information we have about current users, is this something we can just assume, or is there lots of older hardware we will still be asked to support?

Reply:
I have no information about this and would need some form of survey among our users. But I think this would need to be done in addition to some fallback.